### PR TITLE
[Closing] Making template specializations of member functions inline in hpp

### LINKED ISSUE #3839: [compile error] "Cannot add pcl::ColorGradientModality<pcl::PointXYZRGB> as member"
**Issue Summary**
I'm trying to create a wrapper class for `pcl::LINEMOD` which requires me to make `pcl::ColorGradientModality` and `pcl::SurfaceNormalModality` my member variables. I include the following headers and my .h file:

```
#include <pcl/recognition/color_gradient_modality.h>
#include <pcl/recognition/surface_normal_modality.h>
```

but the build fails with the following errors:

```
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `Eigen::internal::binary_evaluator<Eigen::CwiseBinaryOp<Eigen::internal::scalar_quotient_op<float, float>, Eigen::Block<Eigen::Matrix<float, 3, 3, 0, 3, 3>, 1, 3, false> const, Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Matrix<float, 1, 3, 1, 1, 3> const> const>, Eigen::internal::IndexBased, Eigen::internal::IndexBased, float, float>::binary_evaluator(Eigen::CwiseBinaryOp<Eigen::internal::scalar_quotient_op<float, float>, Eigen::Block<Eigen::Matrix<float, 3, 3, 0, 3, 3>, 1, 3, false> const, Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<float>, Eigen::Matrix<float, 1, 3, 1, 1, 3> const> const> const&)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:243: multiple definition of `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneRowDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:243: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneColDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:263: multiple definition of `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneColDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:263: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneRowNonDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:283: multiple definition of `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneRowNonDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:283: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneColNonDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:317: multiple definition of `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneColNonDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:317: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneRowDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:351: multiple definition of `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneRowDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:351: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneColDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:368: multiple definition of `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneColDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:368: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneRowNonDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:385: multiple definition of `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneRowNonDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:385: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneColNonDense(int, int)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:391: multiple definition of `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::convolveOneColNonDense(int, int)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:391: first defined here
CMakeFiles/ubt_detector.dir/src/ubt_detector.cpp.o: In function `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::makeInfinite(pcl::RGB&)':
/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:397: multiple definition of `pcl::filters::Convolution<pcl::RGB, pcl::RGB>::makeInfinite(pcl::RGB&)'
CMakeFiles/ubt_detector.dir/src/main.cpp.o:/usr/local/include/pcl-1.10/pcl/filters/impl/convolution.hpp:397: first defined here
collect2: error: ld returned 1 exit status
```

**To Reproduce**
Create simple class with `pcl::ColorGradientModality` as a member.

**Screenshots/Code snippets/Build information**

My implementation is along the following lines

***MyClass.h***
```
...
#include<pcl/recognition/color_gradient_modality.h>
...

class MyClass
{
public:
    // Member methods
protected:
   ...
   pcl::ColorGradientModality<pcl::PointXYZRGB> color_grad_mod_;
   ...
}
```

***MyClass.cpp***
```
...
MyClass::SomeMethod()
{
    color_grad_mod_.setInputCloud(input_cloud);
    color_grad_mod_.processInputData();
}
...
```

**Your Environment (please complete the following information):**

 - OS: Ubuntu 18.04
 - Compiler: clang6.0.0
 - PCL Version: HEAD
 - PCL Type: Installed/Compiled from source

If PCL was compiled from source or failure in compiling PCL itself:
 - GPU, Kinfu, CUDA enabled? No
 - List and Version of dependencies used: Boost 1.65.1, Eigen3

**Possible Solution**

Based on this [thread](http://www.pcl-users.org/Linking-error-previous-definition-s-pcl-people-HOG-td4026803i20.html) the solutions seems to be to separate `convolution.h` member methods into a separate .cpp file instead of an

```
#include <pcl/recognition/impl/convolution.hpp>
``` 

at the end of the file. Similarly for `color_grad_modality.h` as well.

### PR DESCRIPTION
**Pull request by kunaltyagi**
Fixes #3839

### CODE REVIEW

**Comment by kunaltyagi** [comment]
@truhoang-ubtrobot Could you please confirm that this PR resolves your issue?

**Comment by kunaltyagi** [comment]
Oops. Wrong base branch. Will edit asap.

**Comment by SergioRAgostinho** [comment]
These functions should not be inlined... Are you trying to workaround issues with ODR? You should move them to the cpp files...

**Comment by kunaltyagi** [comment]
`inline` doesn't imply inlined, but yeah, for some larger ones, cpp makes sense.

**Comment by SergioRAgostinho** [comment]
> `inline` doesn't imply inlined, but yeah, for some larger ones, cpp makes sense.

I'm interpreting it as a recommendation to the compiler to the replace the function call and dump the instructions inside the function directly.

> The original intent of the inline keyword was to serve as an indicator to the optimizer that inline substitution of a function is preferred over function call, that is, instead of executing the function call CPU instruction to transfer control to the function body, a copy of the function body is executed without generating the call. This avoids overhead created by the function call (passing the arguments and retrieving the result) but it may result in a larger executable as the code for the function has to be repeated multiple times.
> 
> Since this meaning of the keyword inline is non-binding, compilers are free to use inline substitution for any function that's not marked inline, and are free to generate function calls to any function marked inline. Those optimization choices do not change the rules regarding multiple definitions and shared statics listed above. 

It's just not worth to break the inline semantics to work around the ODR issue.

File: filters/include/pcl/filters/impl/convolution.hpp [hunk 1]
```diff
@@ -239,164 +240,40 @@ namespace pcl
   namespace filters
   {
     template<> pcl::PointXYZRGB
-    Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneRowDense (int i, int j)
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/convolution.hpp:242]: Can you not remove this header entirely? Do the declarations still need to reside here?

File: filters/include/pcl/filters/impl/pyramid.hpp [hunk 2]
```diff
@@ -195,381 +195,20 @@ namespace pcl
...
-                next.at (j,i).b = static_cast<std::uint8_t> (b);
-                next.at (j,i).g = static_cast<std::uint8_t> (g);
-                next.at (j,i).r = static_cast<std::uint8_t> (r);
-              }
-            }
-          }
-        }
-      }    
-    }
+    template <> inline void
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/pyramid.hpp:198]: Not a good candidate for inline.

File: filters/include/pcl/filters/impl/pyramid.hpp [hunk 3]
```diff
@@ -195,381 +195,20 @@ namespace pcl
...
-                next.at (j,i).g = static_cast<std::uint8_t> (g);
-                next.at (j,i).r = static_cast<std::uint8_t> (r);
-                next.at (j,i).a = static_cast<std::uint8_t> (a);
-              }
-            }
-          }
-        }
-      }    
-    }
+    template <> inline void
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/pyramid.hpp:201]: Same comment here.

File: filters/include/pcl/filters/impl/pyramid.hpp [hunk 4]
```diff
@@ -195,381 +195,20 @@ namespace pcl
...
-                next.at (j,i).g = static_cast<std::uint8_t> (g);
-                next.at (j,i).r = static_cast<std::uint8_t> (r);
-              }
-            }
-          }
-        }
-      }    
-    }
-
+    template <> inline void
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/pyramid.hpp:210]: Same comment here.

File: surface/include/pcl/surface/3rdparty/poisson4/bspline_data.hpp [hunk 5]
```diff
@@ -446,7 +446,7 @@ namespace pcl
     {
       POISSON_THROW_EXCEPTION (pcl::poisson::PoissonBadArgumentException, "B-spline up-sampling not supported for degree " << Degree);
     }
-    template<>
+    template<> inline
```
Comment (SergioRAgostinho) [inline surface/include/pcl/surface/3rdparty/poisson4/bspline_data.hpp:449]: Create a cpp file for the inline additions in this file as well. The corresponding CMake file is in
https://github.com/PointCloudLibrary/pcl/blob/f82719543bdb2848ce807bb58d9a98ead1eab5da/surface/CMakeLists.txt#L84-L88

**Comment by SergioRAgostinho** [comment]
I just noticed you we're still working on this. My bad. :eyes:

File: filters/include/pcl/filters/impl/convolution.hpp (hunk 1 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/impl/convolution.hpp:242]: Yes. The declaration is needed to be visible (basically, this is just a normal function now)

File: filters/include/pcl/filters/impl/pyramid.hpp (hunk 2 above)
Comment (kunaltyagi) [reply filters/include/pcl/filters/impl/pyramid.hpp:198]: Yeah, forgot to remove the inline.

File: surface/include/pcl/surface/3rdparty/poisson4/bspline_data.hpp (hunk 5 above)
Comment (kunaltyagi) [reply surface/include/pcl/surface/3rdparty/poisson4/bspline_data.hpp]: Didn't see that. Modified.

**Comment by taketwo** [comment]
CI is red. 

>  'PointXYZRGB' is not a member of 'pcl'

**Comment by kunaltyagi** [comment]
```
[ 17%] Building CXX object filters/CMakeFiles/pcl_filters.dir/src/pyramid.cpp.o
In file included from /__w/1/s/filters/include/pcl/filters/impl/pyramid.hpp:43:0,
                 from /__w/1/s/filters/include/pcl/filters/pyramid.h:170,
                 from /__w/1/s/filters/src/pyramid.cpp:38:
/__w/1/s/common/include/pcl/common/distances.h:180:47: error: 'const struct pcl::RGB' has no member named 'y'
     float diff_x = p2.x - p1.x, diff_y = p2.y - p1.y, diff_z = p2.z - p1.z;
                                               ^
/__w/1/s/common/include/pcl/common/distances.h:180:47: error: 'const struct pcl::RGB' has no member named 'y'
/__w/1/s/common/include/pcl/common/distances.h:180:69: error: 'const struct pcl::RGB' has no member named 'z'
     float diff_x = p2.x - p1.x, diff_y = p2.y - p1.y, diff_z = p2.z - p1.z;
                                                                     ^
/__w/1/s/common/include/pcl/common/distances.h:180:69: error: 'const struct pcl::RGB' has no member named 'z'
```

This error isn't caused by the refactoring. Please suggest how to proceed.

PS: 19.04 CI has gone bonkers, and has starting conflating `std::uintX` with `pcl::uintX` and throwing warnings all around. It also has a few other errors, and I'm still working on removing them

**Comment by SergioRAgostinho** [comment]
> . Please suggest how to proceed.

Start from scratch and let's merge anything that is possible all the way up until the first errors start popping. Basically isolate everything that is causing issues to other more atomic PRs.

**Comment by kunaltyagi** [comment]
> up until the first errors start popping

That'd be the addition of `pyramid.cpp` and including `pyramid.hpp` in it. The error is caused by existing code in [`pyramid.hpp`, L548](https://github.com/PointCloudLibrary/pcl/blob/master/filters/include/pcl/filters/impl/pyramid.hpp#L548)

**Comment by SergioRAgostinho** [comment]
Ok. Revert that part to the original and let's focus on it in a separate PR.

**Comment by kunaltyagi** [comment]
Linking errors in surface:
1. Dependency on filters
    ```
    Linking CXX executable ../bin/pcl_convolve
    CMakeFiles/pcl_convolve.dir/src/convolve.cpp.o:
      In function `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolve_rows_mirror(pcl::PointCloud<pcl::PointXYZRGB>&) [clone ._omp_fn.0]':
      convolve.cpp:(.text._ZN3pcl7filters11ConvolutionINS_11PointXYZRGBES2_E20convolve_rows_mirrorERNS_10PointCloudIS2_EE._omp_fn.0[_ZN3pcl7filters11ConvolutionINS_11PointXYZRGBES2_E12convolveRowsERNS_10PointCloudIS2_EE]+0xb1):
      undefined reference to `pcl::filters::Convolution<pcl::PointXYZRGB, pcl::PointXYZRGB>::convolveOneRowDense(int, int)'
    ...
    ```
2. Dependency within surface
    ````
    Linking CXX shared library ../lib/libpcl_surface.so
    CMakeFiles/pcl_surface.dir/src/3rdparty/poisson4/bspline_data.cpp.o:
      In function `pcl::poisson::Polynomial<0>::BSplineComponent(int)':
      bspline_data.cpp:(.text+0x0): multiple definition of `pcl::poisson::Polynomial<0>::BSplineComponent(int)'
      CMakeFiles/pcl_surface.dir/src/poisson.cpp.o:poisson.cpp:(.text+0x14e0): first defined here
    ````

Tried adding `filters` as subsys_deps in `surface`.

**Comment by kunaltyagi** [comment]
Anyone with CMake skills that can help out here?

**Comment by taketwo** [comment]
I may help, but in its current state, this PR is hard to analyze. There are similar changes going on in several places (filters, surface), then there is formatting change on top. I'd ask you to reduce this PR to a minimal failing changeset. (E.g. only changes in surface.) We'll solve the issues with that, and then port the same approach to the other module.

**Pull request closed without merging**